packages/components/src/text-input/hooks/use-text-input-state.js

```diff
@@ -0,0 +1,494 @@
+	const shiftStep = stepMultiplier * step;
+
+	const isInputTypeNumeric = type === 'number';
+	const isNumeric = isValueNumeric( value );
+
+	const skipAction =
+		! isInputTypeNumeric && ! isNumeric && ! incrementFromNonNumericValue;
+
+	/**
+	 * Create (synced) references to avoid recreating increment and decrement
+	 * callbacks.
+	 */
+	const propRefs = usePropRef( {
+		min,
+		max,
+		value,
+		shiftStep,
+		onChange,
+	} );
+
+	const increment = useCallback(
+		( /** @type {number} */ jumpStep = 0 ) => {
+			if ( skipAction ) return;
+			// TODO: rename variables, since they're in the upper scope?
+			// eslint-disable-next-line no-shadow
+			const { max, min, onChange, shiftStep, value } = propRefs.current;
+			const baseValue = isValueNumeric( value ) ? value : 0;
+			const nextValue = add( jumpStep * step, shiftStep );
+			const next = roundClampString(
+				add( baseValue, nextValue ),
+				min,
+				max,
+				shiftStep
+			);
+
+			onChange( next );
+		},
+		[ skipAction, propRefs, step ]
+	);
+
+	const decrement = useCallback(
+		( /** @type {number} */ jumpStep = 0 ) => {
+			if ( skipAction ) return;
+			// TODO: rename variables, since they're in the upper scope?
+			// eslint-disable-next-line no-shadow
+			const { max, min, onChange, shiftStep, value } = propRefs.current;
+			const baseValue = isValueNumeric( value ) ? value : 0;
+			const nextValue = add( jumpStep * step, shiftStep );
+			const next = roundClampString(
+				subtract( baseValue, nextValue ),
+				min,
+				max,
+				shiftStep
+			);
+
+			onChange( next );
+		},
+		[ skipAction, propRefs, step ]
+	);
+
+	return { increment, decrement };
+}
+
+/**
+ * @typedef UseNumberKeyboardHandlersProps
+ * @property {boolean} [isTypeNumeric] Whether the type is numeric.
+ * @property {boolean} [stopIfEventDefaultPrevented] Stop if event default prevented.
+ * @property {(int: number) => void} [increment] Increment text input number value callback.
+ * @property {(int: number) => void} [decrement] Decrement text input number value callback.
+ */
+
+/**
+ *
+ * @param {UseNumberKeyboardHandlersProps} props
+ * @return {Object} Return object
+ */
+function useNumberKeyboardHandlers( {
+	decrement = noop,
+	increment = noop,
+	isTypeNumeric,
+	stopIfEventDefaultPrevented = true,
+} ) {
+	const keyboardHandlers = useMemo(
+		() => ( {
+			ArrowUp( /** @type {import('react').KeyboardEvent}} */ event ) {
+				if ( ! isTypeNumeric ) return;
+
+				if ( stopIfEventDefaultPrevented && event.isDefaultPrevented() )
+					return;
+
+				event.preventDefault();
+
+				increment();
+			},
+			ArrowDown( /** @type {import('react').KeyboardEvent}} */ event ) {
+				if ( ! isTypeNumeric ) return;
+
+				if ( stopIfEventDefaultPrevented && event.isDefaultPrevented() )
+					return;
+
+				event.preventDefault();
+
+				decrement();
+			},
+		} ),
+		[ decrement, increment, isTypeNumeric, stopIfEventDefaultPrevented ]
+	);
+
+	const handleOnKeyDown = useCallback(
+		( event ) => {
```

Comment:
We'll need to add a type annotation to `event` for TS to correctly infer the return type of the hook.